### PR TITLE
Implement CertifyCreation()

### PR DESCRIPTION
**Pull request by twitchy-jsonp**

### CODE REVIEW

File: tpm2/tpm2.go [hunk 1]
```diff
@@ -1156,6 +1155,42 @@ func Certify(rw io.ReadWriter, parentAuth, ownerAuth string, object, signer tpmu
...
+	params, err := tpmutil.Pack(qualifyingData, creationHash, scheme, tpmutil.RawBytes(tk))
+	if err != nil {
+		return nil, err
+	}
+	return concat(handles, auth, params)
+}
+
+// CertifyCreation generates a signature of a newly-created &
+// loaded TPM object with a signing key signer. Returned values
+// are: attestation data (TPMS_ATTEST), signature, and an error.
```
Comment (awly) [inline tpm2/tpm2.go]: Use named returns instead

File: tpm2/tpm2.go [hunk 2]
```diff
@@ -1156,6 +1155,42 @@ func Certify(rw io.ReadWriter, parentAuth, ownerAuth string, object, signer tpmu
...
+	handles, err := tpmutil.Pack(signer, object)
+	if err != nil {
+		return nil, err
+	}
+	auth, err := encodeAuthArea(HandlePasswordSession, objectAuth)
+	if err != nil {
+		return nil, err
+	}
+	scheme := tpmtSigScheme{AlgRSASSA, AlgSHA256}
+	tk, err := ticket.Encode()
```
Comment (awly) [inline tpm2/tpm2.go]: I just realized that `tpmutil.Pack(ticket)` should "just work" and return the same as `ticket.Encode`.
You can probably pass ticket to `tpmutil.Pack` below directly

File: tpm2/tpm2_test.go [hunk 3]
```diff
@@ -877,3 +877,57 @@ func TestEncodeDecodeCreationData(t *testing.T) {
...
 	}
 }
+
+func TestCreateAndCertifyCreation(t *testing.T) {
+	rw := openTPM(t)
+	defer rw.Close()
+	params := Public{
+		Type:       AlgRSA,
+		NameAlg:    AlgSHA256,
+		Attributes: FlagFixedTPM | FlagFixedParent | FlagSensitiveDataOrigin | FlagRestricted | FlagSign | FlagNoDA | FlagUserWithAuth,
```
Comment (awly) [inline tpm2/tpm2_test.go]: This can be `FlagSignerDefault | FlagNoDA`

File: tpm2/tpm2_test.go [hunk 4]
```diff
@@ -877,3 +877,57 @@ func TestEncodeDecodeCreationData(t *testing.T) {
...
+		AuthPolicy: nil,
```
Comment (awly) [inline tpm2/tpm2_test.go]: skip this field

File: tpm2/tpm2.go [hunk 5]
```diff
@@ -1156,6 +1155,42 @@ func Certify(rw io.ReadWriter, parentAuth, ownerAuth string, object, signer tpmu
...
+func encodeCertifyCreation(objectAuth string, object, signer tpmutil.Handle, qualifyingData, creationHash []byte, ticket *Ticket) ([]byte, error) {
+	handles, err := tpmutil.Pack(signer, object)
+	if err != nil {
+		return nil, err
+	}
+	auth, err := encodeAuthArea(HandlePasswordSession, objectAuth)
+	if err != nil {
+		return nil, err
+	}
+	scheme := tpmtSigScheme{AlgRSASSA, AlgSHA256}
```
Comment (awly) [inline tpm2/tpm2.go]: Pass a `SigScheme` to `CertifyCreation`

File: tpm2/tpm2_test.go [hunk 6]
```diff
@@ -877,3 +877,57 @@ func TestEncodeDecodeCreationData(t *testing.T) {
...
+				Alg:  AlgRSASSA,
+				Hash: AlgSHA256,
+			},
+			KeyBits: 2048,
+			Modulus: big.NewInt(0),
+		},
+	}
+	keyHandle, pub, _, creationHash, tix, _, err := CreatePrimaryEx(rw, HandleEndorsement, pcrSelection, emptyPassword, emptyPassword, params)
+	if err != nil {
+		t.Fatalf("CreatePrimary 2 failed: %s", err)
```
Comment (awly) [inline tpm2/tpm2_test.go]: `CreatePrimaryEx failed`

File: tpm2/tpm2_test.go [hunk 7]
```diff
@@ -877,3 +877,57 @@ func TestEncodeDecodeCreationData(t *testing.T) {
...
+	}
+	defer FlushContext(rw, keyHandle)
+	attestation, signature, err := CertifyCreation(rw, emptyPassword, keyHandle, keyHandle, nil, creationHash, tix)
+	if err != nil {
+		t.Fatalf("CertifyCreation failed: %s", err)
+	}
+	att, err := DecodeAttestationData(attestation)
+	if err != nil {
+		t.Fatalf("DecodeAttestationData(%v) returned error: %v", attestation, err)
```
Comment (awly) [inline tpm2/tpm2_test.go]: keep error messages consistent: `Foo failed: %v`
replace all of: `returned error`, `returned err`, `Signature failed to verify`

File: tpm2/tpm2_test.go [hunk 8]
```diff
@@ -877,3 +877,57 @@ func TestEncodeDecodeCreationData(t *testing.T) {
...
+	}
+	if att.Type != TagAttestCreation {
+		t.Errorf("Expected attestation structure to be of type TagAttestCreation, got %v", att.Type)
```
Comment (awly) [inline tpm2/tpm2_test.go]: got before want

**Comment by twitchy-jsonp** [comment]
PTAL! :smile:

**Pull request merged**